### ISSUE TITLE
Users shown as off-line while they are actually on-line

### ISSUE THREAD

**Issue by ddobrev** [qTox/qTox#2820]
I was on Windows (7 64-bit) using the 64-bit version v1.2-231-gf27214f. A friend was on OS X using version v1.2-184-gad3326b. Over night, I let my computer hibernate instead of turning it off. When I resumed it, my friend appeared to be off-line. I saw in Skype he was actually on-line and asked him - over Skype - what was going on. He said he could not see me either. He had to restart his qTox so that we can see each other again. Just now, after I had only let my computer sleep, not hibernate, the same happened and he had to restart qTox again after I notified him over Skype.

**Comment by tux3** [comment]
This is something that should probably get fixed in toxcore.
What we could do if we want a fix any time soon is force-reconnect if we detect that we just woke up from hibernation.

**Comment by ddobrev** [comment]
A quick fix would be nice indeed, thank you. qTox does have occasional problems but this is the only issue we consider critical because it prevents knowing for sure if the other person is available or not. However, please consider that it's not really a lost connection because both people actually have working qTox instances, it's just that they cannot see each other.

**Comment by LittleVulpix** [comment]
https://github.com/irungentoo/toxcore/issues/1349 ? Although this seems like qtox not reconnecting (automatically) after it disconnects (which it does in sleep / hibernation), so this is a duplicate of https://github.com/tux3/qTox/issues/2542 actually. Close as dupe?

**Comment by ddobrev** [comment]
It does seem like a duplicate but I think this is actually bad news because I can see the original has been filed in May 2015, this is more than half a year for such a critical bug.

**Comment by LittleVulpix** [comment]
@ddobrev I think it stems from the fact that each side (client / core) dev believed the issue to be a problem for the other side! :dancer: 

In reality though, there are two issues.

A) Users not seeing each other online when normally using a tox client -> toxcore issue (core should see people online if they're online, no ifs, no buts)
B) Clients not reconnecting after the device running the client goes into hibernation/sleep and wakes up -> client issue (clients should auto-reconnect unless told otherwise)

**Comment by ddobrev** [comment]
@LittleVulpix thank you for these details. They show me that the problem has been identified so I am now curious about the reason it has not been fixed yet. Is it because the problem cannot be reproduced?

**Comment by LittleVulpix** [comment]
@ddobrev The toxcore issue is a problem of reproduction. It may be a while until it's fixed, but personally I've not seen it happen for quite some time.

The client issue is (or was?) a problem of misunderstanding why this occurs, so I imagine since the creators of qtox know, it could be fixed. But I cannot speak on their behalf. In general though, if the issue is reproducible (which it is, with clear steps) and not too difficult (seems a simple case of "reconnect if connection is lost" or "reconnect if we are waking up from hibernation/sleep" ?), I'd expect the issue to be fixed relatively shortly.

Again though, I'm not a dev, you'll have to take it up with them :)

**Comment by tux3** [comment]
The main problem is that it's not easy to detect when the computer woke up from hibernation, it's a very platform specific thing.
On Windows it's feasible, on OS X it's harder, and on Linux it's a nightmare.

A possible workaround would be to check the time since boot every 10s.
Since we check the time since boot, we're not affected by time adjustments at all.
If there's a jump forward of more than 5 minutes, the system must have gone to sleep and we should restart toxcore (or it's going to stay with broken connections forever).
If there's a jump backwards at all, the clock is severely broken and we should stop checking the time.

On Linux there's CLOCK_BOOTTIME, on Windows there's the system up time in HKEY_PERFORMANCE_DATA, and on OS X I think there's mach_absolute_time().

**Comment by ddobrev** [comment]
@tux3 thank you for your clarifications. In the case of OS X there seem to be reliable and documented ways to detect waking up: https://stackoverflow.com/questions/9402179/reliable-way-to-get-a-machine-wake-user-logged-in-notification-in-cocoa, https://developer.apple.com/library/mac/qa/qa1340/_index.html. I got these through a simple search. Have you seen these approaches? Are there problems with them?

**Comment by tux3** [comment]
@ddobrev I found the same doc, but I have zero experience with Objective C and I also don't have a Mac machine to test, so it would be much harder for me to implement it. 
It's certainly not impossible, though.

**Comment by ddobrev** [comment]
@tux3 I understand. How about Windows, has it been implemented? if not, could you please implement it? The manifestation of the problem I have encountered might as well be because of my (Windows) client, not my friend's.

**Comment by tux3** [comment]
@ddobrev Not yet,someone needs to implement it.

This is actually a dup of #883, but I'll keep this one since it has some more relevant info.

**Comment by capasicum** [comment]
@tux3 @ddobrev 
How about using the code from: [**Listing 3**  Installing an I/O Kit sleep/wake notification](https://developer.apple.com/library/mac/qa/qa1340/_index.html#//apple_ref/doc/uid/DTS10003321-CH1-SOURCECODE3)? 

The source code is in **_C**_, so that should solve it on OS X.

**Comment by ddobrev** [comment]
@capasicum this should work indeed. @tux3 what do you think? I understand it wouldn't solve the problem of you not having a mac but do you think you could perhaps write it in the blind and then have someone with a mac test it for you?

**Comment by tux3** [comment]
@capasicum @ddobrev 
I don't really have the time to work on this bug at the moment, but that could be a solution yes, thanks for finding it.

**Comment by ddobrev** [comment]
@tux3 would you consider giving this more priority? I believe it's a real show stopper not to be able to be certain at any given moment if your contacts are on-line or not. I think this harms many people who decided to try qTox considering that most users do not take the time to complain but rather just move on. I personally would stop using Skype at all if it wasn't for this problem - now I have to every now and then tell people to reconnect because I see them in Skype but not in qTox.

**Comment by tux3** [comment]
I agree that this needs to be fixed, it's certainly pretty high on my list of priorities but I don't have a lot of time to write code at the moment.

**Comment by ddobrev** [comment]
@tux3 I can see you have little time indeed. So I think you could perhaps break this issue up. You could implement it one platform at a time. For example, if you did "just" for Windows, this would critically improve the experience of at least 70 % of all your users. What do you think?

**Comment by ddobrev** [comment]
Fellows, I can definitely see you have a plenty of bugs that need to be worked on but from what I've seen in your list of issues, this one and https://github.com/tux3/qTox/issues/2920 are the only ones that prevent the entire application from functioning at all and happen with any configuration. I understand I might have overlooked other critical problems but my personal impression is that the two mentioned above should get the top priority. I really hope you would either prove me wrong or share my view.

**Comment by ddobrev** [comment]
It's been 4 months since I filed this issue. This is a third of the year, not to mention its previous version at https://github.com/tux3/qTox/issues/883 which is half a year old. There has been neither a workaround nor this supposed fix in toxcore nor the code to reconnect after the computer wakes up. About the latter, I've already said that if it's too much work to do at once, it can be implemented for one OS at a time because having it for just some OS-s won't make it work less well on others, compared to before.
I have renewed hope for qTox since https://github.com/tux3/qTox/issues/2920 was released after all (albeit not for OS X yet). So please fix this one too - it's an universal problem and the fix is going to relieve many users. Also, there are relatively few such universal problems (about 4 or 5 in total) so fixing just one can make a significant impact.

**Comment by zetok** [comment]
> There has been neither a workaround nor this supposed fix in toxcore nor the code to manually reconnect after the computer wakes up.

Err, there is a way to reconnect manually.
`Settings → General → Reconnect`

**Comment by ddobrev** [comment]
No, I meant for the app to reconnect "manually", that is, make an explicit call to reconnect. I have corrected my comment.

**Comment by vladimiry** [comment]
So would it be a solution if app does `Settings → General → Reconnect` action in it's own constantly and in a not resource consuming way. In my opinion a need to do it manually makes app absolutely unusable.

**Comment by ddobrev** [comment]
@vladimiry thank you, this is exactly what I meant in my comment.

**Comment by zetok** [comment]
@vladimiry 

> So would it be a solution if app does `Settings → General → Reconnect` action in it's own constantly and in a not resource consuming way. In my opinion a need to do it manually makes app absolutely unusable.

No, that would be a workaround to toxcore problem, and one that could cause more problems than it fixes.

And not really currently doable without using additional resources.

**Comment by ddobrev** [comment]
@zetok if the real problem is in toxcore, and in addition there are no reasonable workarounds, should we send our comments to https://github.com/irungentoo/toxcore/issues/1349 instead?

**Comment by zetok** [comment]
@ddobrev 
Well, there are 2 problems here: 
- irungentoo/toxcore#1349 is about what @LittleVulpix described as `A)`
- the hibernation/sleep issue, for which toxcore could most likely make some use of a new issue (preferably word it out clearly, otherwise toxcore maintainer might mark it as a duplicate issue)

With that being said, for me e.g. hibernating & restoring from hibernation on ~newest qTox worked just fine on Debian 8 – note though that this was done with a rather small friend list, so it might not be a really valid sample size.

Could be also something OS-specific, dunno.

On a side note,

> (…) and in addition there are no reasonable workarounds (…)

The "reasonable workaround" here is the restart button in settings – while qTox can't exactly fix toxcore's code, nor it does know when toxcore's networking isn't actually working right, it can at least provide a button for the user to restart toxcore.

**Comment by ddobrev** [comment]
@zetok I see. I'll try having https://github.com/irungentoo/toxcore/issues/1349 fixed first.

**Comment by ghost** [comment]
Oh come on, 1 year passed and toxcore doesn't care about this.

## Repost from : https://github.com/irungentoo/toxcore/issues/1349

This issue can be fixable already.
Sinply add these settings into options, and let users use it.

[X] Reconnect Automatically Every (pull down menu V )
1 hours
2 hours
3 hours
6 hours
12 hours
1 day

> if ($qTox_running_time >= 86400) then
> qtox_reconnect(); // Same action when user press proxy 'reconnect' button in qtox settings
> fi

And

[X] Do not reconnect if the user has open chat window
(e.g., if the user using multiple window mode, do not reconnect if there is any chat window opened)

**Comment by ghost** [comment]
https://github.com/irungentoo/toxcore/issues/1349

**Comment by ghost** [comment]
About the tag "upstream": https://github.com/irungentoo/toxcore/issues/1597
So, the main developer disappears(according to reddit), this means "no upstream fix".

qTox's "reconnect interval" is the possible walkaround bugfix I think.

**Comment by LittleVulpix** [comment]
@proxifier the main dev has not disappeared as far as I know. He's just a perfectionist and doesn't want to push working code if it's not working perfectly.

**Comment by ddobrev** [comment]
This continues to heavily plague the user experience. Is either updating toxcore, switching to TokTok/toxcore or completing the refactorings going to eliminate this bug?

**Comment by zetok** [comment]
> Is either updating toxcore, switching to TokTok/toxcore or completing the refactorings going to eliminate this bug?

Will be fixed in toxcore at some point.

**Comment by shalnoff** [comment]
Confirmed. Really actual issue. Simple periodical reconnection (on demand) would solve the problem IMHO.

**Comment by vladimiry** [comment]
Is there some progress? Is there a hope that this long-standing problem will be fixed?

**Comment by zetok** [comment]
> Is there some progress?

Yes and no. Progress happens all the time, not necessarily strictly related to this issue though. There's quite a bit of things in toxcore that need to be done first before there will be focus on this. That is, assuming that this bug won't go away due to refactoring being done.

> Is there a hope that this long-standing problem will be fixed?

At some point, yeah. With that being said, the more people help, the faster things get fixed :)

**Comment by Diadlo** [comment]
Upstream issue fixed, shouldn't be a problem with new toxcore version

**Comment by sudden6** [comment]
Closing since new toxcore version with the fix is out now since a long time.
